# In this file, we pull in postgres data after the collinearity feature reduction adjustment and use Scikit-learn's random forest permutation feature importance method to reduce the features with weak predictive power.  Finally, we run the same Deep Neural Network tests as we did in the baseline tests again to see the new results.

In [16]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import psycopg2
import time

In [17]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import mean_squared_error
from pathlib import Path
import tensorflow as tf
import matplotlib.pyplot as plt

In [18]:
# Download raw data from postgres for stage 1 ETL

conn_string = 'postgres://whnpmxwsiccrtg:53c453893549d2b1e6a4ff92e626a2a08ebcaff66678e50d33e3742f66e3e4f4@ec2-52-4-171-132.compute-1.amazonaws.com/d2ajro4cjr10lb'

db = create_engine(conn_string)
conn = db.connect()

start_time = time.time()
merge1 = pd.read_sql_query('select * from "merged_no_cal"',con=conn)
print("PostGres Download Duration: {} seconds".format(time.time() - start_time))
conn.close ()

PostGres Download Duration: 0.8261852264404297 seconds


In [27]:
# Split our preprocessed data into our features and target arrays
y = merge1["price"]
X = merge1.drop("price",1)

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

# Perform RFR to evaluate feature importance

In [28]:
# Get feature importances using RFR
# import the regressor 
from sklearn.ensemble import RandomForestRegressor

#Instantiate the Decision Tree Regression model:
random_forest_regression_model = RandomForestRegressor(n_estimators = 200, random_state = 78) 
random_forest_regression_model.fit(X_train, y_train)


RandomForestRegressor(n_estimators=200, random_state=78)

In [29]:

# Measuring feature importance using permutation
from sklearn.metrics import r2_score
from sklearn.inspection import permutation_importance


r = permutation_importance(random_forest_regression_model, X_test, y_test, random_state = 78, n_jobs=30)

In [30]:
sorted_idx = r.importances_mean.argsort()
feat_imp_df = pd.DataFrame(data=r.importances_mean[sorted_idx], index = X_test.columns[sorted_idx], columns=['Importance'])
feat_imp_df.sort_values(ascending=False, by='Importance').head(25)

Importance
bathrooms                                          0.326393
cleaning_fee                                       0.136475
accommodates                                       0.123314
host_listings_count                                0.092135
bedrooms                                           0.073703
reviews_per_month                                  0.045878
neighbourhood_cleansed_District 19                 0.033249
security_deposit                                   0.022658
number_of_reviews                                  0.017147
guests_included                                    0.017023
days_host                                          0.013073
availability_365                                   0.012304
cancellation_policy_strict_14_with_grace_period    0.008467
room_type_Entire home/apt                          0.007513
smoke_detector                                     0.006647
minimum_nights                                     0.005648
neighbourhood_cleansed_District 17                 0.005627
free_parking_on_premises                           0.005382
property_type_Loft                                 0.005255
review_scores_rating                               0.005170
elevator                                           0.004835
maximum_nights                                     0.003327
shampoo                                            0.003022
hair_dryer                                         0.002775
lock_on_bedroom_door                               0.002617

In [31]:
cols_to_drop = feat_imp_df[feat_imp_df['Importance'] < 0.015].index

In [32]:
merge2 = merge1.drop(columns=cols_to_drop)

In [33]:
merge2.columns

Index(['host_listings_count', 'accommodates', 'bathrooms', 'bedrooms', 'price',
       'security_deposit', 'cleaning_fee', 'guests_included',
       'number_of_reviews', 'reviews_per_month',
       'neighbourhood_cleansed_District 19'],
      dtype='object')

In [34]:
# Split our preprocessed data into our features and target arrays
y = merge2["price"].values
X = merge2.drop("price",1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [35]:
from sklearn.preprocessing import RobustScaler
scaler = RobustScaler()
# Fit the RobustScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Test 1: Use one input layer and one hidden layer

In [36]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 80
hidden_nodes_layer2 = 30

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 80)                880       
_________________________________________________________________
dense_1 (Dense)              (None, 30)                2430      
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 31        
Total params: 3,341
Trainable params: 3,341
Non-trainable params: 0
_________________________________________________________________


In [19]:
# Compile the model
nn.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

In [20]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
117/117 [==============================] - 1s 1ms/step - loss: 191.4521 - mean_absolute_error: 191.4521
Epoch 2/100
117/117 [==============================] - 0s 896us/step - loss: 126.1267 - mean_absolute_error: 126.1267
Epoch 3/100
117/117 [==============================] - 0s 954us/step - loss: 90.7978 - mean_absolute_error: 90.7978
Epoch 4/100
117/117 [==============================] - 0s 1ms/step - loss: 86.2892 - mean_absolute_error: 86.2892
Epoch 5/100
117/117 [==============================] - 0s 1ms/step - loss: 81.4624 - mean_absolute_error: 81.4624
Epoch 6/100
117/117 [==============================] - 0s 920us/step - loss: 82.2438 - mean_absolute_error: 82.2438
Epoch 7/100
117/117 [==============================] - 0s 943us/step - loss: 75.7267 - mean_absolute_error: 75.7267
Epoch 8/100
117/117 [==============================] - 0s 955us/step - loss: 74.1156 - mean_absolute_error: 74.1156
Epoch 9/100
117/117 [==============================] - 0s 839us/step - los

In [21]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, MAE: {model_accuracy}")

39/39 - 0s - loss: 65.1233 - mean_absolute_error: 65.1233
Loss: 65.123291015625, MAE: 65.123291015625


# Test 2 Add more neurons to each layer

In [22]:
## Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 100
hidden_nodes_layer2 = 50

nn2 = tf.keras.models.Sequential()

# First hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Output layer
nn2.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn2.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 100)               1900      
_________________________________________________________________
dense_4 (Dense)              (None, 50)                5050      
_________________________________________________________________
dense_5 (Dense)              (None, 1)                 51        
Total params: 7,001
Trainable params: 7,001
Non-trainable params: 0
_________________________________________________________________


In [23]:
# Compile the model
nn2.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model2 = nn2.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
117/117 [==============================] - 0s 984us/step - loss: 190.5339 - mean_absolute_error: 190.5339
Epoch 2/50
117/117 [==============================] - 0s 1ms/step - loss: 100.0951 - mean_absolute_error: 100.0951
Epoch 3/50
117/117 [==============================] - 0s 1ms/step - loss: 85.9878 - mean_absolute_error: 85.9878
Epoch 4/50
117/117 [==============================] - 0s 1ms/step - loss: 85.4927 - mean_absolute_error: 85.4927
Epoch 5/50
117/117 [==============================] - 0s 984us/step - loss: 77.6872 - mean_absolute_error: 77.6872
Epoch 6/50
117/117 [==============================] - 0s 923us/step - loss: 75.8674 - mean_absolute_error: 75.8674
Epoch 7/50
117/117 [==============================] - 0s 833us/step - loss: 71.3554 - mean_absolute_error: 71.3554
Epoch 8/50
117/117 [==============================] - 0s 936us/step - loss: 70.0703 - mean_absolute_error: 70.0703
Epoch 9/50
117/117 [==============================] - 0s 823us/step - loss: 70.472

In [24]:
# Evaluate the model using the test data
model_loss2, model_accuracy2 = nn2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss2}")
print(f"Mean Absolute Error: {model_accuracy2}")

39/39 - 0s - loss: 66.3587 - mean_absolute_error: 66.3587
Loss: 66.35865020751953
Mean Absolute Error: 66.35865020751953


# Test 3:  Add an additional layer with 20 neurons

In [25]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 100
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20

nn3 = tf.keras.models.Sequential()

# First hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Output layer
nn3.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn3.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 100)               1900      
_________________________________________________________________
dense_7 (Dense)              (None, 50)                5050      
_________________________________________________________________
dense_8 (Dense)              (None, 20)                1020      
_________________________________________________________________
dense_9 (Dense)              (None, 1)                 21        
Total params: 7,991
Trainable params: 7,991
Non-trainable params: 0
_________________________________________________________________


In [26]:
# Compile the model
nn3.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model3 = nn3.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
117/117 [==============================] - 0s 1ms/step - loss: 187.1350 - mean_absolute_error: 187.1350
Epoch 2/50
117/117 [==============================] - 0s 1ms/step - loss: 93.1732 - mean_absolute_error: 93.1732
Epoch 3/50
117/117 [==============================] - 0s 1ms/step - loss: 84.1385 - mean_absolute_error: 84.1385
Epoch 4/50
117/117 [==============================] - 0s 1ms/step - loss: 77.3480 - mean_absolute_error: 77.3480
Epoch 5/50
117/117 [==============================] - 0s 1ms/step - loss: 74.2579 - mean_absolute_error: 74.2579
Epoch 6/50
117/117 [==============================] - 0s 1ms/step - loss: 71.3518 - mean_absolute_error: 71.3518
Epoch 7/50
117/117 [==============================] - 0s 1ms/step - loss: 73.3742 - mean_absolute_error: 73.3742
Epoch 8/50
117/117 [==============================] - 0s 941us/step - loss: 69.4344 - mean_absolute_error: 69.4344
Epoch 9/50
117/117 [==============================] - 0s 1ms/step - loss: 71.5332 - mean_abs

In [27]:
# Evaluate the model using the test data
model_loss3, model_accuracy3 = nn3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss3}")
print(f"Mean Absolute Error: {model_accuracy3}")

39/39 - 0s - loss: 64.9918 - mean_absolute_error: 64.9918
Loss: 64.99178314208984
Mean Absolute Error: 64.99178314208984


# Test 4:  Add 50 neurons to the first layer and a 4th layer with 10 neurons

In [28]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 150
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10

nn4 = tf.keras.models.Sequential()

# First hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Output layer
nn4.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model

In [29]:
# Compile the model
nn4.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model4 = nn4.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
117/117 [==============================] - 1s 1ms/step - loss: 187.0719 - mean_absolute_error: 187.0719
Epoch 2/50
117/117 [==============================] - 0s 1ms/step - loss: 91.6559 - mean_absolute_error: 91.6559
Epoch 3/50
117/117 [==============================] - 0s 1ms/step - loss: 79.8165 - mean_absolute_error: 79.8165
Epoch 4/50
117/117 [==============================] - 0s 982us/step - loss: 77.2750 - mean_absolute_error: 77.2750
Epoch 5/50
117/117 [==============================] - 0s 1ms/step - loss: 74.4562 - mean_absolute_error: 74.4562
Epoch 6/50
117/117 [==============================] - 0s 1ms/step - loss: 70.9308 - mean_absolute_error: 70.9308
Epoch 7/50
117/117 [==============================] - 0s 1ms/step - loss: 69.6067 - mean_absolute_error: 69.6067
Epoch 8/50
117/117 [==============================] - 0s 1ms/step - loss: 71.3516 - mean_absolute_error: 71.3516
Epoch 9/50
117/117 [==============================] - 0s 946us/step - loss: 70.2134 - mean_a

In [30]:
# Evaluate the model using the test data
model_loss4, model_accuracy4 = nn4.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss4}")
print(f"Mean Absolute Error: {model_accuracy4}")

39/39 - 0s - loss: 64.4242 - mean_absolute_error: 64.4242
Loss: 64.42415618896484
Mean Absolute Error: 64.42415618896484


# Test 5:  Add 50 more neurons to the first layer and add a 5th layer with just 5 neurons

In [31]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 200
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10
hidden_nodes_layer5 = 5

nn5 = tf.keras.models.Sequential()

# First hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Fifth hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer5,kernel_initializer='normal', activation="relu"))


# Output layer
nn5.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn5.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 200)               3800      
_________________________________________________________________
dense_16 (Dense)             (None, 50)                10050     
_________________________________________________________________
dense_17 (Dense)             (None, 20)                1020      
_________________________________________________________________
dense_18 (Dense)             (None, 10)                210       
_________________________________________________________________
dense_19 (Dense)             (None, 5)                 55        
_________________________________________________________________
dense_20 (Dense)             (None, 1)                 6         
Total params: 15,141
Trainable params: 15,141
Non-trainable params: 0
__________________________________________________

In [32]:
# Compile the model
nn5.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model5 = nn5.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
117/117 [==============================] - 1s 1ms/step - loss: 192.1460 - mean_absolute_error: 192.1460
Epoch 2/50
117/117 [==============================] - 0s 1ms/step - loss: 94.5303 - mean_absolute_error: 94.5303
Epoch 3/50
117/117 [==============================] - 0s 1ms/step - loss: 80.2050 - mean_absolute_error: 80.2050
Epoch 4/50
117/117 [==============================] - 0s 1ms/step - loss: 74.1033 - mean_absolute_error: 74.1033
Epoch 5/50
117/117 [==============================] - 0s 1ms/step - loss: 68.3855 - mean_absolute_error: 68.3855
Epoch 6/50
117/117 [==============================] - 0s 1ms/step - loss: 70.9930 - mean_absolute_error: 70.9930
Epoch 7/50
117/117 [==============================] - 0s 1ms/step - loss: 68.7326 - mean_absolute_error: 68.7326
Epoch 8/50
117/117 [==============================] - 0s 1ms/step - loss: 68.7847 - mean_absolute_error: 68.7847
Epoch 9/50
117/117 [==============================] - 0s 1ms/step - loss: 69.3293 - mean_absol

In [33]:
# Evaluate the model using the test data
model_loss5, model_accuracy5 = nn5.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss5}")
print(f"Mean Absolute Error: {model_accuracy5}")

39/39 - 0s - loss: 64.8520 - mean_absolute_error: 64.8520
Loss: 64.85202026367188
Mean Absolute Error: 64.85202026367188


# Test 6:  Add 150 neurons to first layer and change second activation function to tanh

In [34]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 350
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10
hidden_nodes_layer5 = 5

nn6 = tf.keras.models.Sequential()

# First hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="tanh"))

# Third hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Fifth hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer5,kernel_initializer='normal', activation="relu"))

# Output layer
nn6.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn6.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_21 (Dense)             (None, 350)               6650      
_________________________________________________________________
dense_22 (Dense)             (None, 50)                17550     
_________________________________________________________________
dense_23 (Dense)             (None, 20)                1020      
_________________________________________________________________
dense_24 (Dense)             (None, 10)                210       
_________________________________________________________________
dense_25 (Dense)             (None, 5)                 55        
_________________________________________________________________
dense_26 (Dense)             (None, 1)                 6         
Total params: 25,491
Trainable params: 25,491
Non-trainable params: 0
__________________________________________________

In [35]:
# Compile the model
nn6.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model6 = nn6.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
117/117 [==============================] - 1s 1ms/step - loss: 204.0117 - mean_absolute_error: 204.0117
Epoch 2/50
117/117 [==============================] - 0s 1ms/step - loss: 125.9113 - mean_absolute_error: 125.9113
Epoch 3/50
117/117 [==============================] - 0s 1ms/step - loss: 100.8778 - mean_absolute_error: 100.8778
Epoch 4/50
117/117 [==============================] - 0s 1ms/step - loss: 80.9416 - mean_absolute_error: 80.9416
Epoch 5/50
117/117 [==============================] - 0s 1ms/step - loss: 70.8528 - mean_absolute_error: 70.8528
Epoch 6/50
117/117 [==============================] - 0s 1ms/step - loss: 70.2353 - mean_absolute_error: 70.2353
Epoch 7/50
117/117 [==============================] - 0s 1ms/step - loss: 69.6332 - mean_absolute_error: 69.6332
Epoch 8/50
117/117 [==============================] - 0s 1ms/step - loss: 68.6942 - mean_absolute_error: 68.6942
Epoch 9/50
117/117 [==============================] - 0s 1ms/step - loss: 65.1020 - mean_a

In [36]:
# Evaluate the model using the test data
model_loss6, model_accuracy6 = nn6.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss6}")
print(f"Mean Absolute Error: {model_accuracy6}")

39/39 - 0s - loss: 59.1812 - mean_absolute_error: 59.1812
Loss: 59.18115997314453
Mean Absolute Error: 59.18115997314453


# From all these tests, the best performing Deep Neural Network (the last test), was off, on average, by about $59.18